### ISSUE TITLE
Get rid of wasm-gc

### ISSUE THREAD

**Issue by pepyakin** [paritytech/substrate#1262]
[wasm-gc](https://github.com/alexcrichton/wasm-gc) got deprecated but we still use it in our build pipeline.

The author mentions that:

> The Rust compiler now natively supports --gc-sections when linking wasm executables, which means wasm executables already have 90% of their garbage removed when coming out of the compiler.

But for us this is not entirely true and the size of the runtime is:

- 882K baseline
- 636K with wasm-gc

And

> The wasm-pack (and wasm-bindgen) project will already run this by default for you, so there's no need to run it again.

But we don't use neither of these.

We need to figure out what do we want to use.

**Comment by pepyakin** [comment]
According to `twiggy diff node_runtime.wasm node_runtime.compact.wasm`:

```
 Delta Bytes │ Item
─────────────┼─────────────────────────────────
     -252601 ┊ <total>
      -68707 ┊ custom section '.debug_str'
      -56447 ┊ custom section '.debug_line'
      -52165 ┊ custom section '.debug_info'
      -16040 ┊ custom section '.debug_ranges'
      -10261 ┊ custom section '.debug_pubnames'
       -6072 ┊ custom section '.debug_pubtypes'
       -1432 ┊ custom section '.debug_abbrev'
         -20 ┊ "function names" subsection
         -18 ┊ custom section '.debug_macinfo'
          -3 ┊ __wasm_call_ctors
          -1 ┊ elem[0]
          -1 ┊ func[357]
```

so major source of the cruft is `debug_` sections.

There is a tool called `wasm-strip` in the wabt suite, which strips all custom sections.

```
882K baseline
643K after wasm-strip
636K after wasm-gc
```

however, if we decide to depend on wabt in our build pipeline we could then use wabt's wasm-opt, which could optimize binary (including pruning of dead symbols) and produce a binary with size of  `561K`.

**Comment by bkchr** [comment]
Why do we have debug sections in a release build? Shouldn't they be removed?

**Comment by pepyakin** [comment]
I'd say they should be removed! But for now this is not the case (and it has been this way for a while).
And I'm not sure why is this that way, but I asked why.

**Comment by pepyakin** [comment]
So, a quick recap: 

- substrate uses a wasm file to encode logic of blockchain. We call it wasm runtime
- an example wasm runtime is written in rust, built with cargo. Rust uses LLVM LLD linker to link rust libraries and produce a final wasm blob
- by default, rustc puts a quite some junk into the final binary and that junk is unnecessary.
- to strip this unnecessary stuff, we use `wasm-gc`. 

There is a chance that since the last time we checked the things have changed. Notably, there is a LLD flag `-s` which can be used to strip all unnecessary sections.

To check if that actually helps we need:

1. Disable wasm-gc. We can comment it on the line below just for the experiment.
    https://github.com/paritytech/substrate/blob/7010ec7716e0edf97d61a29bd0c337648b3a57ae/core/utils/wasm-builder/src/wasm_project.rs#L354-L360

1. Try build the wasm runtime passing this flag to LLD and disable wasm-gc. This can be achieved by building substrate with the following command:
    ```
    WASM_BUILD_RUSTFLAGS='-C link-arg=-s' cargo build --release
    ```

You should be able to find the runtime files under `target/release/wbuild` directory. Keep in mind that if you skip wasm-gc then there won't be the `.compact.wasm`.

If the results turn out to be satisfying, we can then get rid of wasm-gc enitrely.

**Comment by kianenigma** [comment]
without wasm-gc

```
TODO
```

with wasm-gc

```bash
// compact 
→ pwd
/Users/kianenigma/Desktop/Parity/substrate/target/release/wbuild/node-runtime
 → du -sk node_runtime.compact.wasm 
1156	node_runtime.compact.wasm

// Original
→ pwd                      
.../substrate/target/release/wbuild/target/wasm32-unknown-unknown/release
→ du -sk node_runtime.wasm        
1480	node_runtime.wasm
```

**Comment by Swader** [comment]
Just a note that the test suite depends on wasm-gc and the prerequisites installation script doesn't pull it in. So to reduce new user friction, it should either be added to that one, or removed from the tests imo.

**Comment by bkchr** [comment]
Which script? The init script installs it.
Nevertheless, it is documented in the readme and having clear install instructions is always better than having any scripts that work on one distribution.

**Comment by Swader** [comment]
You're right, I've been relying too much on the getsubstrate.io script which does not pull it in, but the README does document it.

**Comment by bkchr** [comment]
@shawntabrizi ^^^

**Comment by bkchr** [comment]
(That is also a script I really would like to get rid of!)

**Comment by Swader** [comment]
I ran into it from the Substrate Kitties workshop and [here](https://substrate.dev/docs/en/next/tutorials/start-a-private-network-with-substrate).

As a tech ed I'm all for simplification and abstracting bootstrapping from the user, but I do agree that if we cannot provide a coherent experience to newcomers, such a script shouldn't be a thing. Maybe something like a vagrant box with all the prerequisites installed would be better.

**Comment by shawntabrizi** [comment]
We can and should update the `getsubstrate.io` script for the time being.

@bkchr I am 100% with you, let's get rid of all the scripts. For now though, I don't see a good solution for installing all these prerequisites.

Maybe the best thing is to literally have an install page with each operating system and instructions?

@Swader vagrant is a virtual env?

**Comment by shawntabrizi** [comment]
Here is a PR to update the `getsubstrate.io` script for now: https://github.com/paritytech/scripts/pull/101

**Comment by Swader** [comment]
@shawntabrizi vagrant is a VM tool for composing and running reproducible headless VMs of little size and resource intensity. For a good while I was maintaining https://github.com/Swader/homestead_improved

Basically it lets you pre-install stuff with provisioning scripts (shell, ansible, etc.) into a headless distro like a 200MB ubuntu, similar to what you might find in WSL, but it's not raw - it's all the prereqs in place. So no downloading rust, running updates, etc. The organization / author keeps the box up to date, the users merely run vagrant box update once in a while, though this is optional because once they run an instance of the VM they can just update it as any other headless VM.

Here's a better explained rundown of mine from ages ago: https://www.sitepoint.com/re-introducing-vagrant-right-way-start-php/ - mentions PHP but has nothing to do with it specifically.

In a nutshell, all OSes then have the _exact same thing_. On windows, it works exactly the same as on OS X or Linux. So suddenly all of your users have to follow exactly one set of instructions which always works (download vagrant, oraclevm and this box) and they have the same env from which to report bugs, which is coincidentally identical to the env on which you can then reproduce them.

**Comment by kianenigma** [comment]
@Swader if you are recommending us maintaining vagrant boxes (or similar IaC tools) that'd be cool. We already have docker builds that can be used mostly to play with the ui and less to hack substrate itself. 

But, such things would never be enough. We need manual installation as well, always.

**Comment by Swader** [comment]
I agree. Let me see if I can put together a box, I could use one myself so I'm sure I could dedicate part of my time to maintaining it. We can evaluate once done.

**Comment by DrSensor** [comment]
Hi, is it okay if I work on this issue?

**Comment by bkchr** [comment]
You can work on this, but our experiments show that you probably will not find anything that achieves the same result.

**Comment by shawntabrizi** [comment]
@pepyakin this should be closed right?

**Comment by pepyakin** [comment]
I wasn't following for what is happening on the wasm-builder side lately, but just a quick glance revealed that it still does invoke `wasm-gc`. So apparently it is still being used.

**Comment by shawntabrizi** [comment]
How do we go about getting rid of it?

If we can't, should this issue be closed?

**Comment by pepyakin** [comment]
We should check how littered the binaries produced by the rustc are nowadays. I would be pleasantly surprised if it improved but I don't hold my breath.

The alternative is to use wasm-opt. I expect it would give way better results. I historically wasn't so sure about it but now I guess less so. FWIW, the contract team uses it for preparing contracts by default.

**Comment by expenses** [comment]
I've started looking into this. [`binaryen-rs`](https://github.com/pepyakin/binaryen-rs) unfortunately has a very long build time due to compiling binaryen via gcc. It might be possible to use a pre-installed version of binaryen though. @pepyakin do you know if that's do-able?

**Comment by expenses** [comment]
The difference isn't huge imo:
```
du before_compact.wasm after_compact.wasm
2928	before_compact.wasm
2700	after_compact.wasm
```
```
twiggy diff before_compact.wasm after_compact.wasm
 Delta Bytes │ Item
─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     -304100 ┊ data[0]
     -142011 ┊ "function names" subsection
     -101888 ┊ <node_runtime::Call as frame_support::traits::dispatch::UnfilteredDispatchable>::dispatch_bypass_filter::h7d3fd923cf8fd14e
      +94582 ┊ code[256]
      -63778 ┊ node_runtime::_::<impl parity_scale_codec::codec::Decode for node_runtime::Call>::decode::h307a9f9b65f4fda8
      +61535 ┊ code[107]
      -48851 ┊ <pallet_assets::pallet::Call<T,I> as frame_support::traits::dispatch::UnfilteredDispatchable>::dispatch_bypass_filter::hc22452fada11af84
      -47899 ┊ node_runtime::_::<impl parity_scale_codec::codec::Encode for node_runtime::Event>::encode_to::h969c3ff99399b607
      -47226 ┊ node_runtime::_::<impl parity_scale_codec::codec::Decode for node_runtime::Call>::decode::ha00c8e5063e970e8
      -45630 ┊ node_runtime::_::<impl parity_scale_codec::codec::Encode for node_runtime::Call>::encode_to::hf70ff279a2fc7797
      +45271 ┊ code[101]
      +45069 ┊ code[262]
      +43329 ┊ code[164]
      +40773 ┊ custom section '.debug_str'
      -39285 ┊ node_runtime::_::<impl parity_scale_codec::codec::Decode for node_runtime::Event>::decode::h0895f1dad2dd60d6
      -38797 ┊ <node_runtime::NposCompactSolution16 as sp_npos_elections::CompactSolution>::into_assignment::h36f919769403d82e
      +38166 ┊ code[214]
      +37295 ┊ code[143]
      -37156 ┊ <pallet_democracy::Call<T> as frame_support::traits::dispatch::UnfilteredDispatchable>::dispatch_bypass_filter::hfd7c1543aff6391d
      -36219 ┊ schnorrkel::vrf::VRFOutput::attach_input_hash::h539b2808f611581e
     +311808 ┊ ... and 5099 more.
     -235012 ┊ Σ [5119 Total Rows]
```

**You also get the best size reduction by running `wasm-opt -Oz` on the `wasm_gc`'d binary:**

```
du before_compact.wasm after_compact.wasm wasm_gc_then_wasm_opt.wasm
2928	before_compact.wasm
2700	after_compact.wasm
2600	wasm_gc_then_wasm_opt.wasm
```

so TLDR; wasm optimisation is still a dark art.

**Comment by expenses** [comment]
Anyway; here's a branch for that: https://github.com/paritytech/substrate/compare/ashley-wasm-builder-wasm-opt

**Comment by pepyakin** [comment]
Can you add to comparison just plain `wasm-opt` with -Oz (and perhaps with -O3, or what's the analogue of that?) but without running wasm-gc before that? 

wasm-gc performs a really simple tree shaking pass. I am sure that wasm-opt is capable of doing that as well.

Regarding binaryen/gcc: I wasn't aware of that actually. Most testing happened on macOS and clang. Sometimes I do use gcc but it worked fast enough that I didn't notice anything. We can use locally installed wasm-opt. That, however, will somehow complicate deterministic builds. `wasm-opt` will be another variable needed to be fixed during such a build.

**Comment by expenses** [comment]
> Can you add to comparison just plain `wasm-opt` with -Oz (and perhaps with -O3, or what's the analogue of that?) but without running wasm-gc before that?

Here's a full comparison:

```
du *
3180	node_runtime.wasm
3024	node_runtime_gc.wasm
2708	node_runtime_gc_O3.wasm
2720	node_runtime_gc_O4.wasm
2708	node_runtime_gc_Oz.wasm
2708	node_runtime_O3.wasm
2720	node_runtime_O4.wasm
2708	node_runtime_Oz.wasm
```

So it seems that `wasm-gc` doesn't manage to reduce the sizes down to what I got above, only the `wasm_gc` function call? Strange.

**Comment by pepyakin** [comment]
😰  I have realized one important thing

wasm-opt butchers DWARF

Meaning that if we start relying on wasm-opt we will make on-chain wasm execution introspection and debugging even harder.  (No issue/writeup yet to point, but things like onchain alarms, onchain logging without overhead, debuggers (forward and reverse)). Now, to get a trace you will have to have to re-execute the original pre-wasm-opt binary.

I think this should be considered before embracing wasm-opt in the build.

**Comment by olanod** [comment]
Seems like WASM supports [`external_debug_info`](https://yurydelendik.github.io/webassembly-dwarf/#external-DWARF) is it something that would help? more importantly are any of the tools able to generate that and make use of it?

**Comment by pepyakin** [comment]
The problem I was talking is that tools like wasm-opt when they move code around during optimizations do not patch the DWARF data. That means be it external or embedded in the wasm file it will be broken anyway.

**Comment by kianenigma** [comment]
One way or another this should be removed from the Runtime project.

**Comment by nazar-pc** [comment]
Copied from https://github.com/paritytech/substrate/pull/10704#issuecomment-1017523486:

Maybe it is worth removing compact version entirely?
I just tested `-C lto=fat -C codegen-units=1` and got roughly the same size for regular build as compact version prior to this. It is also a bit faster in that benchmark.

Of course running wasm-opt on top helps even more, but that is something every chain maintainer can apply on top before doing runtime upgrade, right?

**Comment by pepyakin** [comment]
Yes, they can. I would go further and say that the users opting into wasm-opt would be ideal.

`wasm-gc` just removes unreachable symbols. OTOH, `wasm-opt` performs more elaborate transformations. It enables improvements in code size. However it also brings risks of some transformations not being sound. IMO the final user should evaluate pros and cons whether they want to take on the risk for the extra wins of code size and take into account other considerations as the DWARF point I brought up above.

As you, @nazar-pc, mentioned in https://github.com/paritytech/substrate/pull/10704#issue-1109078800 `wasm-gc` does not support newest wasm features like simd128. Since we eventually want to enable them, https://github.com/paritytech/substrate/issues/10707, to reap the performance benefits that's a sign that `wasm-gc` deprecation goes from OK to an active nuisance. Some maybe it's finally time to get rid of it. I think it maybe OK to just skip the post-build optimization steps and just rely on compression for now. Then, optionally, we could add `wasm-opt` as an opt-in.

**Comment by nazar-pc** [comment]
I can prepare a PR with `wasm-gc` removed and LTO/codegen units exposed as methods on `WasmBuilder` (so `build.rs` on runtime examples will be able to use them to compensate for wasm size increase) if you'd like.

**Comment by nazar-pc** [comment]
Numbers below are for `node-runtime` on NVMe SSD with AMD 5900X 12C24T processor.

Status quo:
* baseline size 4_360_748 bytes
* compact size (wasm-gc) 3_974_726 bytes
* compact+compress size (wasm-gc+zstd) 1_030_234 bytes
* `wasm-opt -Os -O2` on top of baseline 3_670_017 bytes
* release build time from scratch is 1m 46s
  * out of it `node-runtime`'s build script is 72.3s

`-C lto=fat -C codegen-units=1`:
* uncompressed size 3_980_524 bytes
* compressed size (zstd) 1_036_487 bytes
* `wasm-opt -Os -O2` on top of uncompressed 3_374_880 bytes
* release build time from scratch is 2m 23s
  * out of it `node-runtime`'s build script is 110.3s

**Comment by pepyakin** [comment]
Thanks for the thorough reply!

The build time metric is a bit daunting. I wonder though if that can be a result of `codgen-units=1`. I heard anecdotes that cgu number does not influence the quality of the generated code in LTO mode lately. Thus I can ask you to run the LTO case without `-Ccodegen-units=1`

**Comment by nazar-pc** [comment]
Size without `codgen-units=1` barely changes.

`-C lto=fat`:
* uncompressed size 4_345_542 bytes
* compressed size (zstd) 1_092_160 bytes
* `wasm-opt -Os -O2` on top of uncompressed 3_644_745 bytes
* release build time from scratch is 1m 42s
  * out of it `node-runtime`'s build script is 69.8s

**Comment by athei** [comment]
I don't get it. AFAIK wasm builder builds the runtime with `lto` in release profile already. How to do you see those differences with only `lto`enabled:
https://github.com/paritytech/substrate/blob/48e0b415b94d8ce14e0ebbbdb813263afa27f02e/utils/wasm-builder/src/wasm_project.rs#L231

**Comment by koute** [comment]
> I don't get it. AFAIK wasm builder builds the runtime with `lto` in release profile already. How to do you see those differences with only `lto` enabled:

Just guessing - probably because of the extra `codegen-units=1`, and also AFAIK by default `lto` runs in `thinlto` mode, doesn't it? (So it'd be different than `lto=fat`.)

> The build time metric is a bit daunting.

Hmm.... indeed, although extra 40s doesn't seem *that* bad considering how long everything else takes to build.

Maybe we should have separate cargo profiles for a proper release build and a "development" release build?

**Comment by pepyakin** [comment]
> Maybe we should have separate cargo profiles for a proper release build and a "development" release build?

Yes, I was thinking about something along these lines too. Alternatively, maybe finally get the debug builds (with lots of opt-level overrides for performance critical crates).

> Hmm.... indeed, although extra 40s doesn't seem that bad considering how long everything else takes to build.

The problem is that it's 40 seconds for one runtime. In [polkadot](https://github.com/paritytech/polkadot) we have 4 ༼ ༎ຶ ෴ ༎ຶ༽

**Comment by koute** [comment]
> 
> > Hmm.... indeed, although extra 40s doesn't seem that bad considering how long everything else takes to build.
> 
> The problem is that it's 40 seconds for one runtime. In [polkadot](https://github.com/paritytech/polkadot) we have 4 ༼ ༎ຶ ෴ ༎ຶ༽

Yes but, wouldn't they be compiled *in parallel*?

**Comment by athei** [comment]
> Just guessing - probably because of the extra codegen-units=1

But @nazar-pc reports a difference for `lto="fat"` **only**. This is what confuses me.

> and also AFAIK by default lto runs in thinlto mode, doesn't it? (So it'd be different than lto=fat.)

Not according to [this documentation](https://doc.rust-lang.org/cargo/reference/profiles.html#lto). AFAIK `true == fat`.

**Comment by pepyakin** [comment]
> Yes but, wouldn't they be compiled in parallel?

Oh, right, that's true

**Comment by nazar-pc** [comment]
> But @nazar-pc reports a difference for lto="fat" only. This is what confuses me.

I think this is caused by incremental compilation, I did not wipe `target` in between those initial tests until I got to measure time.

> Yes but, wouldn't they be compiled in parallel?

Assuming you have enough CPU cores of course :wink:

**Comment by athei** [comment]
> I think this is caused by incremental compilation, I did not wipe target in between those initial tests until I got to measure time.

You also reported different code sizes.

**Comment by nazar-pc** [comment]
> You also reported different code sizes.

I just re-ran it a few times and I do reproducibly get different sizes with and without `.append_to_rust_flags("-C lto=fat")` added to `WasmBuilder` call.

I don't think incremental compilation in Rust guarantees for builds from scratch to be fully identical to some incremental rebuilds, but this is not it regardless.

**Comment by nazar-pc** [comment]
I checked commands that are being generated, and with `release_profile.insert("lto".into(), true.into());` you only get `-C lto`, which explains the difference.

**Comment by koute** [comment]
> > Yes but, wouldn't they be compiled in parallel?
> 
> Assuming you have enough CPU cores of course wink

Well, indeed I do, whole 32 of them. (:

Anyhow, this seems like it's something we should measure. Maybe you could also compile `polkadot` (so all of the runtimes get compiled) while limiting the compilation to a certain number of cores (on Linux you should be able to do this through `taskset`) and compare how the compile time changes assuming we have N cores?

**Comment by nazar-pc** [comment]
Polkadot compilation on the same machine as above, also attaching timing results.

Status quo:
```
   Completed rococo-runtime v0.9.13 build script (run) in 143.0s
   Completed westend-runtime v0.9.13 build script (run) in 148.2s
   Completed polkadot-runtime v0.9.13 build script (run) in 150.6s
   Completed kusama-runtime v0.9.13 build script (run) in 151.9s
...
    Finished release [optimized] target(s) in 6m 15s
```
[cargo-timing-before.html.zip](https://github.com/paritytech/substrate/files/7914359/cargo-timing-before.html.zip)

`-C lto=fat -C codegen-units=1`:
```
   Completed rococo-runtime v0.9.13 build script (run) in 169.4s
   Completed westend-runtime v0.9.13 build script (run) in 187.0s
   Completed polkadot-runtime v0.9.13 build script (run) in 193.7s
   Completed kusama-runtime v0.9.13 build script (run) in 212.5s
...
    Finished release [optimized] target(s) in 7m 06s
```
[cargo-timing-after.html.zip](https://github.com/paritytech/substrate/files/7914702/cargo-timing-after.html.zip)

Runtime compilation into WASM is a single-threaded process, so with 4+ cores time overhead doesn't add up and compilation happens in parallel.

**Comment by bkchr** [comment]
We could just remove wasm-gc and do the optimal build only for when we prepare a release. We already do this for stuff like `log` where we disable it completlty for the on chain build to reduce the final binary size.

**Comment by athei** [comment]
> We could just remove wasm-gc and do the optimal build only for when we prepare a release. We already do this for stuff like `log` where we disable it completely for the on chain build to reduce the final binary size.

https://github.com/paritytech/substrate/pull/10747

However, I kept `wasm-gc` around for now as it still reduces the file size. `wasm-opt` needs to be applied first but I didn't want to conflate that in one PR.

**Comment by bkchr** [comment]
> could also look into compiling the runtime with `opt-level = z`.

I have done this 2 years ago or whatever and this wasn't bringing any good results. I think it was even slower.

**Comment by athei** [comment]
> > could also look into compiling the runtime with `opt-level = z`.
> 
> I have done this 2 years ago or whatever and this wasn't bringing any good results. I think it was even slower.

This isn't surprising. It optimizes for smaller code rather than "fast code". Often that isn't necessarily slower. I am not sure how important a smaller runtime would be and if it makes sense to sacrifice some performance `opt-level = s` could be a good middle ground.

**Comment by ggwpez** [comment]
> This isn't surprising. It optimizes for smaller code rather than "fast code". Often that isn't necessarily slower. I am not sure how important a smaller runtime would be and if it makes sense to sacrifice some performance. `opt-level = s` could be a good middle ground.

There are cases where the performance improves since the smaller binary causes less cache misses when fetching the instructions. Would have to test it on the exact hardware.  
Rust also seems to have [guided Optimizations](https://doc.rust-lang.org/1.41.1/rustc/profile-guided-optimization.html) to get some automatic cache optimization.  
But I'm not sure if this is actually a good idea, since we try to get worst-cast performance and the Guided Optimizations makes some assumptions about the normal program flow.  
This could lead to good benchmarking performance but worse performance when we hit a branch that was not automatically optimized.

**Comment by bkchr** [comment]
IMO we should not over-complicate it. Just keep the compiler optimizing for performance, remove stuff in the final binary that we don't need and then we compress it anyway again.

**Comment by nazar-pc** [comment]
`-C lto=fat -C codegen-units=1` is already used in `production` profile, does that mean there is nothing preventing `wasm-gc` from being removed?

**Comment by athei** [comment]
I think it sill reduces the code size a bit. Didn't check by how much, though.

**Comment by nazar-pc** [comment]
The difference right now is roughly this:
* zstd `polkadot_runtime.wasm` 1_182_538 bytes
* zstd `polkadot_runtime.compact.wasm` 1_120_459 bytes

So it is only ~5.2% increase, which is significant, but not deal-breaker, especially considering that things got smaller after `-C lto=fat -C codegen-units=1` introduction.

**Comment by pepyakin** [comment]
i wonder where this difference comes from?

can you run `twiggy diff` over those two?

**Comment by nazar-pc** [comment]
Uncompressed size:
* `polkadot_runtime.wasm` is 4_954_045 bytes
* `polkadot_runtime.compact.wasm` is 4_527_986 bytes

```
$ twiggy diff polkadot_runtime.wasm polkadot_runtime.compact.wasm 
 Delta Bytes │ Item
─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     +300764 ┊ "function names" subsection
      -51119 ┊ custom section '.debug_str'
      -23698 ┊ custom section '.debug_info'
      -22742 ┊ custom section '.debug_line'
      -16368 ┊ custom section '.debug_pubnames'
       -8873 ┊ <polkadot_runtime::Call as frame_support::traits::dispatch::UnfilteredDispatchable>::dispatch_bypass_filter::hb866e76c7c626a49
       -7840 ┊ custom section '.debug_ranges'
       -7772 ┊ polkadot_runtime::Runtime::metadata::h169c0e3519f6dcf9
       -3199 ┊ polkadot_runtime::_::<impl parity_scale_codec::codec::Encode for polkadot_runtime::Call>::encode_to::h4d530ea869712805
       -3018 ┊ <xcm_executor::XcmExecutor<Config> as xcm::v2::traits::ExecuteXcm<<Config as xcm_executor::config::Config>::Call>>::execute_xcm_in_credit::h3f8e6465ee293b4b
       -2832 ┊ <sp_runtime::generic::era::Era as scale_info::TypeInfo>::type_info::h1a1dae6524123429
       -2685 ┊ polkadot_runtime::_::<impl parity_scale_codec::codec::Encode for polkadot_runtime::Event>::encode_to::h783a8bffe709a93e
       -2395 ┊ <(TupleElement0,TupleElement1) as frame_support::traits::hooks::OnRuntimeUpgrade>::on_runtime_upgrade::he126fdfea7233115
       -2321 ┊ polkadot_runtime_parachains::initializer::<impl polkadot_runtime_parachains::initializer::pallet::Pallet<T>>::apply_new_session::hdd9528bf14999f39
       -2194 ┊ <pallet_staking::pallet::pallet::Call<T> as frame_support::traits::dispatch::UnfilteredDispatchable>::dispatch_bypass_filter::h1f225e22d27b0cf4
       -2128 ┊ pallet_session::<impl pallet_session::pallet::Pallet<T>>::rotate_session::h4ec98104757f4ffe
       -1841 ┊ <xcm::v1::multilocation::MultiLocation as core::convert::TryFrom<xcm::v0::multi_location::MultiLocation>>::try_from::h7793c69f1b61502b
       -1817 ┊ custom section '.debug_abbrev'
       -1731 ┊ pallet_election_provider_multi_phase::unsigned::<impl pallet_election_provider_multi_phase::pallet::Pallet<T>>::prepare_election_result::he31bdc622f13337b
       -1435 ┊ <(TupleElement0,TupleElement1) as frame_support::traits::hooks::OnInitialize<BlockNumber>>::on_initialize::hf592aba116c5ad05
     -200248 ┊ ... and 2740 more.
     -426059 ┊ Σ [2760 Total Rows]
```
